nexus/src/db/datastore/volume.rs

```diff
@@ -447,6 +467,195 @@ impl DataStore {
+    // to temp_volume_id.
+    //
+    // As this is part of a saga, it will be able to handle being replayed
+    // If we call this twice, any work done the first time through should
+    // not happen again, or be undone.
+    pub async fn volume_remove_rop(
+        &self,
+        volume_id: Uuid,
+        temp_volume_id: Uuid,
+    ) -> Result<bool, Error> {
+        #[derive(Debug, thiserror::Error)]
+        enum RemoveReadOnlyParentError {
+            #[error("Error removing read only parent: {0}")]
+            DieselError(#[from] diesel::result::Error),
+
+            #[error("Serde error removing read only parent: {0}")]
+            SerdeError(#[from] serde_json::Error),
+
+            #[error("Database in unexpected state")]
+            UnexpectedDatabaseState,
+        }
+        type TxnError = TransactionError<RemoveReadOnlyParentError>;
+
+        // In this single transaction:
+        // - Get the given volume from the volume_id from the database
+        // - Extract the volume.data into a VolumeConstructionRequest (VCR)
+        // - Create a new VCR, copying over anything from the original VCR,
+        //   but, replacing the read_only_parent with None.
+        // - Put the new VCR into volume.data, then update the volume in the
+        //   database.
+        // - Get the given volume from temp_volume_id from the database
+        // - Extract the temp volume.data into a VCR
+        // - Create a new VCR, copying over anything from the original VCR,
+        //   but, replacing the read_only_parent with the read_only_parent
+        //   data from original volume_id.
+        // - Put the new temp VCR into the temp volume.data, update the
+        //   temp_volume in the database.
+        self.pool()
+            .transaction(move |conn| {
+                // Grab the volume in question. If the volume record was already
+                // deleted then we can just return.
+                let volume = {
+                    use db::schema::volume::dsl;
+
+                    let volume = dsl::volume
+                        .filter(dsl::id.eq(volume_id))
+                        .select(Volume::as_select())
+                        .get_result(conn)
+                        .optional()?;
+
+                    let volume = if let Some(v) = volume {
+                        v
+                    } else {
+                        // the volume does not exist, nothing to do.
+                        return Ok(false);
+                    };
+
+                    if volume.time_deleted.is_some() {
+                        // this volume is deleted, so let whatever is deleting
+                        // it clean it up.
+                        return Ok(false);
+                    } else {
+                        // A volume record exists, and was not deleted, we
+                        // can attempt to remove its read_only_parent.
+                        volume
+                    }
```

Comment:
For external API resources, we declare them in `nexus/src/db/lookup.rs`, and then you can replace much of this with something like:

```
let (.., db_volume) = LookupPath(opctx, self)
    .volume_id(volume_id)
    .optional_fetch()
    .await?;
```

That will make the query for you (L516-520), ignore anything where `time_deleted IS NOT NULL` (L529-532), etc.  It'll also do an authz check.

This would be predicated on doing the authz work I mentioned above (because you need an `OpContext` to do this).  Again, I'm not sure if this is worth doing.  In the limit that's what I'd expect here but I'm not sure if there's a reason we don't seem to use the `LookupPath` API for other internal resources.

Reply:
Going to thank the review for their good suggestion, then consider implementing it at some later time.